File: components/camel-scala/src/test/scala/org/apache/camel/scala/dsl/ResequencerTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.camel.scala.dsl

import org.junit.Test
import builder.RouteBuilder

/**
 * Test case for message resequencer
 */
class ResequencerTest extends ScalaTestSupport {
  
  @Test
  def testSimpleResequencer() {
    "mock:a" expect { _.received("message 1", "message 2", "message 3", "message 4", "message 5") } 
    test {
      "direct:a" ! ("message 5", "message 1", "message 3", "message 2", "message 4")
    }
  }
  
  @Test
  def testBlockResequencer() {
    "mock:b" expect (_.received("message 5", "message 1", "message 3", "message 2", "message 4"))
    "mock:c" expect (_.received("message 1", "message 2", "message 3", "message 4", "message 5"))
    test {
      "direct:b" ! ("message 5", "message 1", "message 3", "message 2", "message 4")
    }
  }
  
  @Test
  def testBatchResequencer() {
    "mock:d" expect (_.received("message 5", "message 1", "message 3", "message 2"))
    "mock:e" expect (_.count = 0)
    test {
      "direct:d" ! ("message 5", "message 1", "message 3", "message 2")
    }
    "mock:d" expect (_.count = 5)
    "mock:e" expect (_.received("message 1", "message 2", "message 3", "message 4", "message 5"))
    test {
      "direct:d" ! "message 4"
    }
  }
    
  val builder =
    new RouteBuilder {
       //START SNIPPET: simple
       "direct:a" resequence (_.in) to "mock:a"
       //END SNIPPET: simple
       
       //START SNIPPET: block
       "direct:b" ==> {
         to ("mock:b")
         resequence (_.in) {
           to ("mock:c")
         }
       }
       //END SNIPPET: block
       
       //START SNIPPET: batch
       "direct:d" ==> {
         to ("mock:d")
         resequence(_.in).batch(5) {
           to ("mock:e")
         }
       }
       //END SNIPPET: batch
    }

}

File: core/camel-core/src/test/java/org/apache/camel/processor/ResequencerTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.camel.processor;

import java.util.List;

import org.apache.camel.ContextTestSupport;
import org.apache.camel.Endpoint;
import org.apache.camel.Route;
import org.apache.camel.builder.RouteBuilder;
import org.apache.camel.component.mock.MockEndpoint;
import org.apache.camel.impl.engine.DefaultChannel;
import org.apache.camel.impl.engine.DefaultRoute;
import org.apache.camel.processor.errorhandler.DefaultErrorHandler;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ResequencerTest extends ContextTestSupport {
    protected Endpoint startEndpoint;
    protected MockEndpoint resultEndpoint;

    @Test
    public void testSendMessagesInWrongOrderButReceiveThemInCorrectOrder() throws Exception {
        resultEndpoint.expectedBodiesReceived("Guillaume", "Hiram", "James", "Rob");
        sendBodies("direct:start", "Rob", "Hiram", "Guillaume", "James");
        resultEndpoint.assertIsSatisfied();
    }

    @Test
    public void testRestartRoute() throws Exception {
        resultEndpoint.expectedBodiesReceived("Guillaume", "Hiram", "James", "Rob");
        sendBodies("direct:start", "Rob", "Hiram", "Guillaume", "James");
        resultEndpoint.assertIsSatisfied();

        context.getRouteController().stopRoute("myRoute");

        // wait just a little bit
        Thread.sleep(5);
        resultEndpoint.reset();

        context.getRouteController().startRoute("myRoute");

        resultEndpoint.expectedBodiesReceived("Donald", "Goofy", "Jack");
        sendBodies("direct:start", "Jack", "Donald", "Goofy");
        resultEndpoint.assertIsSatisfied();
    }

    @Override
    @BeforeEach
    public void setUp() throws Exception {
        super.setUp();
        resultEndpoint = getMockEndpoint("mock:result");
    }

    @Override
    @AfterEach
    public void tearDown() throws Exception {
        super.tearDown();
    }

    @Override
    protected boolean useJmx() {
        // use jmx only when running the following test(s)
        return getName().equals("testBatchResequencerTypeWithJmx");
    }

    @Override
    protected RouteBuilder createRouteBuilder() {
        return new RouteBuilder() {
            public void configure() {
                // START SNIPPET: example
                from("direct:start").routeId("myRoute")
                        .resequence().body().timeout(50).to("mock:result");
                // END SNIPPET: example
            }
        };
    }

    @Test
    public void testBatchResequencerTypeWithJmx() throws Exception {
        testBatchResequencerTypeWithoutJmx();
    }

    @Test
    public void testBatchResequencerTypeWithoutJmx() throws Exception {
        List<Route> list = getRouteList(createRouteBuilder());
        assertEquals(1, list.size(), "Number of routes created: " + list);

        Route route = list.get(0);
        DefaultRoute consumerRoute = assertIsInstanceOf(DefaultRoute.class, route);

        DefaultChannel channel = assertIsInstanceOf(DefaultChannel.class, unwrapChannel(consumerRoute.getProcessor()));

        assertIsInstanceOf(DefaultErrorHandler.class, channel.getErrorHandler());

        assertIsInstanceOf(Resequencer.class, channel.getNextProcessor());
    }

}
